### PR TITLE
HPCC-29467 STD library access to current query elapsed time

### PR DESCRIPTION
**Pull request by afishbeck**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: plugins/logging/logging.cpp [hunk 1]
```diff
@@ -21,8 +21,9 @@
 #include <ctype.h>
 #include "logging.hpp"
 #include "jlog.hpp"
+#include "roxiehelper.ipp"
 
-#define LOGGING_VERSION "LOGGING 1.0.1"
+#define LOGGING_VERSION "LOGGING 1.0.2"
 static const char * compatibleVersions[] = {
```
Comment (afishbeck) [inline plugins/logging/logging.cpp:27]: @ghalliday I'm not entirely sure how the versioning works, can you comment?

**Comment by afishbeck** [comment]
@ghalliday changed target to 8.12.x at users request.  Open to discussion.

File: plugins/logging/logging.cpp (hunk 1 above)
Comment (richardkchapman) [reply plugins/logging/logging.cpp:27]: We used to use versioning on the plugins to support mixed systems and plugins that could be installed just on the eclserver and then get pushed automatically to other nodes when they needed them. We haven't supported that for a long time, and the version usage has thus rather fallen into disuse. We don't normally change the version string at all any more AFAIK

Comment (ghalliday) [reply plugins/logging/logging.cpp:27]: Yes, no need to change the version number.

File: plugins/logging/CMakeLists.txt [hunk 2]
```diff
@@ -32,6 +32,8 @@ include_directories (
          ./../../system/include 
          ./../../system/jlib 
          ./../../rtl/include
+         ./../../rtl/eclrtl
+         ./../../common/thorhelper
```
Comment (ghalliday) [inline plugins/logging/CMakeLists.txt:36]: I don't like adding this dependency - (or the dynamic cast).  Logically could this functionality apply to any engine?

File: plugins/logging/logging.cpp [hunk 3]
```diff
@@ -111,3 +113,11 @@ LOGGING_API char * LOGGING_CALL logGenerateGloballyUniqueId()
     appendGloballyUniqueId(ret);
     return ret.detach();
 }
+
+LOGGING_API unsigned int LOGGING_CALL logGetRoxieElapsedMs(ICodeContext *ctx)
+{
+    const IRoxieContextLogger *roxieLogger = dynamic_cast<const IRoxieContextLogger*>(&ctx->queryContextLogger());
```
Comment (ghalliday) [inline plugins/logging/logging.cpp:119]: The cleanest would be to extend either ICodeContext or IContextLogger

**Review by ghalliday — COMMENTED** [review]
Not sure about the structure.

File: plugins/logging/CMakeLists.txt (hunk 2 above)
Comment (afishbeck) [reply plugins/logging/CMakeLists.txt:36]: I had considered that, but wasn't as sure of equivalent meaning in other components.  I suppose basing it on workunit start time would be most equivalent to the roxie start of query?  I would then make the roxie running from workunit queue version based on that as well.

**Comment by afishbeck** [comment]
@ghalliday refactored as suggested.  Squashed to make refactoring clear.

File: ecl/eclagent/eclagent.ipp [hunk 4]
```diff
@@ -512,6 +512,7 @@ public:
     virtual IUserDescriptor *queryUserDescriptor();
     virtual void selectCluster(const char * cluster);
     virtual void restoreCluster();
+    virtual unsigned getElapsedMs() const override {UNIMPLEMENTED;} //Implementing for roxie first, not sure what this means from an eclagent perspective.  ECL plugin implies elapsed time of entire query
```
Comment (afishbeck) [inline ecl/eclagent/eclagent.ipp:515]: I've created https://track.hpccsystems.com/browse/HPCC-29490 to consider implementing for thor/hthor.

**Comment by afishbeck** [comment]
@ghalliday also the extra dependencies are now gone.  Moving the method to the ICodeContext interface did trigger the eclhelper interface check.

**Review by ghalliday — APPROVED** [review]
@afishbeck The change looks good, but we may need to target master.  We don't normally change ICodeContext in a point release - especially a previous version.  I'll discuss with Richard.

**Pull request merged**